nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+      'symbol': 'NEO',
+      'amount': '10',
+    },
+    {
+      'assetID': '602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7',
+      'symbol': 'GAS',
+      'amount': '777.0001',
+    },
+    {
+      'assetID': 'c36aee199dbba6c3f439983657558cfb67629599',
+      'symbol': 'NKN',
+      'amount': '0.00000233',
+    },
+    {
+      'assetID': 'fc732edee1efdf968c23c20a9628eaa5a6ccb934',
+      'symbol': 'NNC',
+      'amount': '2000',
+    }
+  ],
+  'AbKNY45nRDy6B65YPVz1B6YXiTnzRqU2uQ': [
+    {
+      'assetID': '1578103c13e39df15d0d29826d957e85d770d8c9',
+      'symbol': 'PHX',
+      'amount': '11000',
+    }
+  ]
+}
+</pre>
+
+'''Provider Request Handling'''
+
+Upon receiving this request, fetch the latest balance(s) for each given account, for the specific asset if applicable.
+Typechecking will be required on the input argument, as it can be eather a single BalanceRequest object, or an array of BalanceRequest objects. In the case where a specific asset is not provided to check the balance for, the wallet provider will fetch balances for all assets and tokens for that account.
+
+
+====getStorage====
+Reads the raw value in smart contract storage.
+
+
+'''Method Interface'''
+
+<pre>
+function getStorage({ scriptHash: string, key: string, network?: string }): Promise<StorageResponse>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface GetStorageArgs {
+  scriptHash: string; // script hash of the smart contract to invoke a read on
+  key: string; // key of the storage value to retrieve from the contract
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+}
+</pre>
+
+
+'''Success return value'''
+
+<pre>
+interface StorageResponse {
+  result: string; // The raw value that's stored in the contract
+}
+</pre>
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_REFUSED`|`RPC_ERROR`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+getStorage({
+  scriptHash: '505663a29d83663a838eee091249abd167e928f5',
+  key: 'game.status'
+})
+.then(res => {
+  const value = res.result;
+  console.log('Storage value: ' + value);
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_REFUSED:
+      console.log('Connection dApp not connected. Please call the "connect" function.');
+      break;
+    case RPC_ERROR:
+      console.log('There was an error when broadcasting this transaction to the network.');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Validate and format the inputs
+** Script hash is valid
+* Broadcast the RPC request
+* Return the RPC response results
+
+Example
+<pre>
+{
+  result: 'hello world'
```

Comment:
This could cause problems on some languages, if using string to represent non-visible chars. a `hex` return option would be nice.

Reply:
this is just an error in example. Yes, they will return hexstrings back from the contract, but tried to show it as a human readable string in the example. Will update.